<p style="align: center;"><img src="https://static.tildacdn.com/tild6636-3531-4239-b465-376364646465/Deep_Learning_School.png", width=600, height=300></p>

<h3 style="text-align: center;"><b>Физтех-Школа Прикладной математики и информатики (ФПМИ) МФТИ</b></h3>

<h2 style="text-align: center;"><b>Библиотека <a href="http://pandas.pydata.org/">pandas</a></b></h2>

Библиотека `pandas` активно используется в современном data science для работы с данными, которые могут быть представлены в виде таблиц (а это очень, очень большая часть данных)

`pandas` есть в пакете Anaconda, но если вдруг у Вас её по каким-то причинам нет, то можно установить, раскомментировав следующую команду:

In [0]:
from google.colab import drive
drive.mount('/content/drive')
%cd 'drive'
%cd 'My Drive'
%cd 'projects'
%cd 'deep_learning_school'
%cd '1.3'

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive
/content/drive
/content/drive/My Drive
/content/drive/My Drive/projects
/content/drive/My Drive/projects/deep_learning_school
/content/drive/My Drive/projects/deep_learning_school/1.3


In [0]:
#!pip install pandas

In [0]:
import numpy as np
import pandas as pd

Нужны данные. Про AppStore слышали? :)

### Скачать данные: https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps

Нас интересует файл `AppStore.csv`.

Кстати, `.csv` (Comma Separated Values) - наверное, самый частый формат данных в современном data science. По сути, это то же самое, что и `.xls` (и `.xlsx`), то есть таблица. 

Итак, *чтение файла с данными*:

In [0]:
import os
os.listdir()

['Copy of [seminar]matplotlib.ipynb',
 'Copy of [seminar,prodv]2_linalg.ipynb',
 'Copy of [seminar,prodv]2_Numpy.ipynb',
 'app-store-apple-data-set-10k-apps',
 'Copy of [seminar]pandas.ipynb']

In [0]:
data = pd.read_csv('./app-store-apple-data-set-10k-apps/AppleStore.csv')

Посмотрим, что же такое переменная `data`:

In [0]:
data

Unnamed: 0          id  ... lang.num  vpp_lic
0              1   281656475  ...       10        1
1              2   281796108  ...       23        1
2              3   281940292  ...        3        1
3              4   282614216  ...        9        1
4              5   282935706  ...       45        1
...          ...         ...  ...      ...      ...
7192       11081  1187617475  ...        1        1
7193       11082  1187682390  ...        1        1
7194       11087  1187779532  ...        1        1
7195       11089  1187838770  ...        2        1
7196       11097  1188375727  ...        2        1

[7197 rows x 17 columns]

In [0]:
type(data)

pandas.core.frame.DataFrame

Тип == `pandas.core.frame.DataFrame`, обычно говорят просто "датафрейм", то есть "кусок данных".

Можно вывести информацию по датафрейму:

In [0]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7197 entries, 0 to 7196
Data columns (total 17 columns):
Unnamed: 0          7197 non-null int64
id                  7197 non-null int64
track_name          7197 non-null object
size_bytes          7197 non-null int64
currency            7197 non-null object
price               7197 non-null float64
rating_count_tot    7197 non-null int64
rating_count_ver    7197 non-null int64
user_rating         7197 non-null float64
user_rating_ver     7197 non-null float64
ver                 7197 non-null object
cont_rating         7197 non-null object
prime_genre         7197 non-null object
sup_devices.num     7197 non-null int64
ipadSc_urls.num     7197 non-null int64
lang.num            7197 non-null int64
vpp_lic             7197 non-null int64
dtypes: float64(3), int64(9), object(5)
memory usage: 956.0+ KB


И более статистически значимую информацию:

In [0]:
data.describe()

Unnamed: 0            id  ...     lang.num      vpp_lic
count   7197.000000  7.197000e+03  ...  7197.000000  7197.000000
mean    4759.069612  8.631310e+08  ...     5.434903     0.993053
std     3093.625213  2.712368e+08  ...     7.919593     0.083066
min        1.000000  2.816565e+08  ...     0.000000     0.000000
25%     2090.000000  6.000937e+08  ...     1.000000     1.000000
50%     4380.000000  9.781482e+08  ...     1.000000     1.000000
75%     7223.000000  1.082310e+09  ...     8.000000     1.000000
max    11097.000000  1.188376e+09  ...    75.000000     1.000000

[8 rows x 12 columns]

Итак, наши данные -- это какая-то **информация про приложения в AppStore**. Давайте посмотрим, что тут есть:

In [0]:
# все столбцы датафрейма (названия)
data.columns

Index(['Unnamed: 0', 'id', 'track_name', 'size_bytes', 'currency', 'price',
       'rating_count_tot', 'rating_count_ver', 'user_rating',
       'user_rating_ver', 'ver', 'cont_rating', 'prime_genre',
       'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'],
      dtype='object')

---

#### Задание 0

О чём данные? (судя по столбцам)

---

Все значения датафрейма:

In [0]:
# одна строка == описание одного объекта, в данном случае -- приложения из AppStore
data.values

array([[1, 281656475, 'PAC-MAN Premium', ..., 5, 10, 1],
       [2, 281796108, 'Evernote - stay organized', ..., 5, 23, 1],
       [3, 281940292, 'WeatherBug - Local Weather, Radar, Maps, Alerts',
        ..., 5, 3, 1],
       ...,
       [11087, 1187779532, 'Bret Michaels Emojis + Lyric Keyboard', ...,
        1, 1, 1],
       [11089, 1187838770, 'VR Roller Coaster World - Virtual Reality',
        ..., 0, 2, 1],
       [11097, 1188375727, 'Escape the Sweet Shop Series', ..., 0, 2, 1]],
      dtype=object)

In [0]:
data.values[0]

array([1, 281656475, 'PAC-MAN Premium', 100788224, 'USD', 3.99, 21292, 26,
       4.0, 4.5, '6.3.5', '4+', 'Games', 38, 5, 10, 1], dtype=object)

In [0]:
data.shape

(7197, 17)

-- то есть 7197 строк и 17 столбцов.

Важно уметь обращаться к определённой строке или столбцу dataframe'а. Индексирование очень похоже на numpy, однако есть свои тонкости:

* Получить весь столбец (в даном случае столбец 'track_name'):

In [0]:
data['track_name']

0                                         PAC-MAN Premium
1                               Evernote - stay organized
2         WeatherBug - Local Weather, Radar, Maps, Alerts
3       eBay: Best App to Buy, Sell, Save! Online Shop...
4                                                   Bible
5                                        Shanghai Mahjong
6                  PayPal - Send and request money safely
7                                 Pandora - Music & Radio
8                             PCalc - The Best Calculator
9                                             Ms. PAC-MAN
10                              Solitaire by MobilityWare
11                                       SCRABBLE Premium
12                   Google – Search made just for mobile
13                       Bank of America - Mobile Banking
14                                               FreeCell
15                 TripAdvisor Hotels Flights Restaurants
16                                               Facebook
17         Yel

In [0]:
type(data['track_name'])

pandas.core.series.Series

In [0]:
data['track_name'][0]

'PAC-MAN Premium'

`pandas.core.series.Series` -- это тип подвыборки датафрейма. С этим типом работают реже, но стоит знать, что он есть.

* Получить определённую строку:

In [0]:
data.iloc[10]

Unnamed: 0                                 11
id                                  284791396
track_name          Solitaire by MobilityWare
size_bytes                           49618944
currency                                  USD
price                                    4.99
rating_count_tot                        76720
rating_count_ver                         4017
user_rating                               4.5
user_rating_ver                           4.5
ver                                    4.10.1
cont_rating                                4+
prime_genre                             Games
sup_devices.num                            38
ipadSc_urls.num                             4
lang.num                                   11
vpp_lic                                     1
Name: 10, dtype: object

Более продвинутое и полезное индексирование:

In [0]:
data.iloc[5, 10]

'1.8'

-- `data.iloc[i, j]`, где `i` -- номер строки, `j` -- номер столбца

Почти все типы индексирования:

* к строкам по числовому индексу, к столбцам -- по имени

In [0]:
data.loc[[0, 1, 2], ['track_name', 'id']]

track_name         id
0                                  PAC-MAN Premium  281656475
1                        Evernote - stay organized  281796108
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292

* к строкам и к столбцам по числовому индексу

In [0]:
data.iloc[[1, 3, 5], [2, 1]]

track_name         id
1                          Evernote - stay organized  281796108
3  eBay: Best App to Buy, Sell, Save! Online Shop...  282614216
5                                   Shanghai Mahjong  283619399

* поддерживает и то, и то, но **запрещён**:

In [0]:
data.ix[[2,1,0], ['track_name', 'id']]

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


track_name         id
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292
1                        Evernote - stay organized  281796108
0                                  PAC-MAN Premium  281656475

In [0]:
data.ix[[0,1,2], [2, 1]]

c:\users\izakharkin\anaconda3\envs\vision\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


track_name         id
0                                  PAC-MAN Premium  281656475
1                        Evernote - stay organized  281796108
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292

Точно также дело обстоит со **срезами**:

In [0]:
data.iloc[5:100, 0:5]

Unnamed: 0         id  ... size_bytes  currency
5            6  283619399  ...   10485713       USD
6            7  283646709  ...  227795968       USD
7            8  284035177  ...  130242560       USD
8            9  284666222  ...   49250304       USD
9           10  284736660  ...   70023168       USD
..         ...        ...  ...        ...       ...
95         108  306169895  ...  168192000       USD
96         109  306257910  ...   17485824       USD
97         110  306375551  ...    5234688       USD
98         111  306468004  ...  237913088       USD
99         112  306561234  ...  233513984       USD

[95 rows x 5 columns]

Но обычно нужно уметь отвечать на более осмысленные вопросы, например:

* сколько приложений имеют рейтинг юзеров >= 4?

In [0]:
len(data[data['user_rating'] >= 4])

4781

* какие есть валюты в столбце currency?

In [0]:
np.unique(data['currency'])

array(['USD'], dtype=object)

* сколько приложений жанра Games? А тех, которые Games и рейтинг >= 4?

In [0]:
# 1
len(data[data['prime_genre'] == 'Games'])

3862

In [0]:
# 2
len(data[(data['user_rating'] >= 4) & (data['prime_genre'] == 'Games')])

2868

Оператор "ИЛИ" аналогично &, только он через |

In [0]:
len(data[(data['user_rating'] >= 4) | (data['prime_genre'] == 'Games')])

5775

---

#### Задание 1

Получите ответ на следующие вопросы (с помощью pandas):

1). Сколько всего жанров и какие они?  
2). Сколько суммарно байт весят все приложения жанра Finance?  
3). Сколько всего бесплатных приложений, у которых рейтинг юзеров больше трёх?  
4). Есть ли приложения, которые стоят больше 100 долларов?

In [0]:
data.columns

Index(['Unnamed: 0', 'id', 'track_name', 'size_bytes', 'currency', 'price',
       'rating_count_tot', 'rating_count_ver', 'user_rating',
       'user_rating_ver', 'ver', 'cont_rating', 'prime_genre',
       'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'],
      dtype='object')

In [0]:
# Ваш код здесь
print(len(np.unique(data['prime_genre'])))
print(np.unique(data['prime_genre']))
print(sum((data['price'] == 0) & (data['user_rating'] > 3)))
print((data['price'] > 0).any())

23
['Book' 'Business' 'Catalogs' 'Education' 'Entertainment' 'Finance'
 'Food & Drink' 'Games' 'Health & Fitness' 'Lifestyle' 'Medical' 'Music'
 'Navigation' 'News' 'Photo & Video' 'Productivity' 'Reference' 'Shopping'
 'Social Networking' 'Sports' 'Travel' 'Utilities' 'Weather']
2944
True


---

Часто надо сделать что-то такое:

In [0]:
import numpy as np

In [0]:
np.array(data['size_bytes'])

array([100788224, 158578688, 100524032, ..., 111322112,  97235968,
        90898432])

То есть взять столбец из таблицы и засунуть его в np.array(). Это можно сделать и просто обратившись к .values, ведь его тип и есть np.array:

In [0]:
type(data['size_bytes'].values)

numpy.ndarray

Больше про pandas можно найти по этом полезным ссылкам:

* Официальные туториалы: http://pandas.pydata.org/pandas-docs/stable/tutorials.html

* Статья на Хабре от OpenDataScience сообщества: https://habr.com/company/ods/blog/322626/

* Подробный гайд: https://media.readthedocs.org/pdf/pandasguide/latest/pandasguide.pdf

Главное в работе с новыми библиотеками -- не бояться тыкать в разные функции, смотреть типы возвращаемых объектов и активно пользоваться Яндексом, а ещё лучше понимать всё из docstring'а (`Shift+Tab` при нахождении курсора внутри скобок функции).